Fix/adapt epa parser

Bachibouzouk opened the pull request:
Fix #605, fix #606, fix #3 

Here are the steps I would follow 

- [x] Reproduced the benchmark test ABE_Grid_PV_battery via EPA interface, here is the json [epa_benchmark.json.zip](https://github.com/rl-institut/multi-vector-simulator/files/5591993/epa_benchmark.json.zip)

- [x] Conversion to MVS 
Use the following script to run the conversion (*after* having pulled this branch obviously)
```
import json
from multi_vector_simulator.utils.data_parser import convert_epa_params_to_mvs
with open("epa_benchmark.json") as json_file:
    epa = json.load(json_file)
    
dict_values = convert_epa_params_to_mvs(epa_dict)

with open("epa_benchmark_converted_to_mvs.json", "w") as json_file:
    json.dump(dict_values, json_file, indent=4)
```

- [x] After adding manually above fields, run the MVS online (mvs-eland.rl-institut.de) with the eap_benchmark.json
(this can also be done locally with  `mvs_tool -i epa_benchmark_converted_to_mvs.json -o MVS_tests_benchmarks -f -png` if there are to many errors)
- [x] Compare the outputs (png graphs, network graphs etc) with the outputs from the mvs command
```
mvs_tool -i tests/benchmark_test_inputs/ABE_grid_PV_battery/ -ext csv -o MVS_tests_benchmarks -f -png

```
**Changes proposed in this pull request**:

The following steps were realized, as well (if applies):
- [ ] Use in-line comments to explain your code
- [ ] Write docstrings to your code ([example docstring](https://multi-vector-simulator.readthedocs.io/en/latest/Developing.html#format-of-docstrings))
- [ ] For new functionalities: Explain in readthedocs
- [x] Write test(s) for your new patch of code (pytests, assertion debug messages)
- [x] Update the CHANGELOG.md
- [x] Apply black (`black . --exclude docs/`)
- [x] Check if benchmark tests pass locally (`EXECUTE_TESTS_ON=master pytest`)



<sub>*For more information on how to contribute check the [CONTRIBUTING.md](https://github.com/rl-institut/multi-vector-simulator/blob/dev/CONTRIBUTING.md).*<sub>

Bachibouzouk commented on src/multi_vector_simulator/utils/constants.py:
```diff
@@ -180,7 +173,7 @@
 # be in a required parameters dict
 REQUIRED_JSON_PARAMETERS = {
-    ECONOMIC_DATA: [CURR, DISCOUNTFACTOR, LABEL, PROJECT_DURATION, TAX],
+    ECONOMIC_DATA: [CURR, DISCOUNTFACTOR, PROJECT_DURATION, TAX],
```
```suggestion
    ECONOMIC_DATA: [CURR, DISCOUNTFACTORS, PROJECT_DURATION, TAX],
```

Bachibouzouk commented on src/multi_vector_simulator/utils/constants.py:
```diff
@@ -191,22 +184,14 @@
-        STORE_OEMOF_RESULTS,
-        TIMESTEP,
-    ],
+    SIMULATION_SETTINGS: [EVALUATED_PERIOD, OUTPUT_LP_FILE, START_DATE, TIMESTEP,],
```
```suggestion
    SIMULATION_SETTINGS: [EVALUATED_PERIOD, OUTPUT_LP_FILE, START_DATE, TIMESTEP,],
    JKHsdkashdkaj
```

smartie2076 commented:
I am taking over this PR now. I will document what I did here.

1. Download `epa_benchmark.json`
2. Create function to translate the EPA json to a MVS json (`utils.helpers.translates_epa_strings_to_mvs_readable()`), output:
```
>>> import multi_vector_simulator.utils.helpers as helpers
>>> helpers.translates_epa_strings_to_mvs_readable("./epa_benchmark", "epa_benchmark.json")
WARNING:root:The parameters group 'constraints' is not present in the EPA parameters to be parsed into MVS json format
WARNING:root:The parameters group 'fixcost' is not present in the EPA parameters to be parsed into MVS json format
```
3. Added following entries manually into the converted json file (Both of these fields should be provided by the EPA!):
```
    "constraints": {
        "minimal_renewable_factor": {
            "unit": "factor",
            "value": 0.3
        }
    },
    "fixcosts": {},
```

smartie2076 commented:
With all of the above changes, a valid input file for the MVS can be created from the current EPA output (as per the data that @Bachibouzouk posted his EPA output first):

[mvs_config.json.txt](https://github.com/rl-institut/multi-vector-simulator/files/5690505/mvs_config.json.txt)

The report can be generated:

[simulation_report.pdf](https://github.com/rl-institut/multi-vector-simulator/files/5690508/simulation_report.pdf)

As well as the results:

[results_and_logfile.zip](https://github.com/rl-institut/multi-vector-simulator/files/5690513/results_and_logfile.zip)

It is a simple scenario with following energy system:
![energy_system_graph](https://user-images.githubusercontent.com/44204527/102114266-4addb080-3e3a-11eb-9f20-e54c7a15866c.png)

smartie2076 commented:
Left to do: 

- [ ] Decide which adaptations are to be automatized
- [ ] Automatize adaptations
- [ ] Replace string represenatations of EPA directly (not working)

Bachibouzouk commented:
@smartie2076 - the potential problem is that if EPA change and their outputed json change, then we need to redo a lot of manual steps. It also requires manual change to the assets fields and rebuilding the model everytime (this is why I ended up making a very simple benchmark) with drag and drop before we can generate the json, this is a lengthy process. I think it is better to make only the non-negotiable changes (such as unit field as a unitless value does not make physical sense) on EPA side and provide flexibility with the EPA-MVS and MVS-EPA parsers for the other formatting/parameters (like a parameter which should be an int and is provided as a float is relatively easy to cover on our side; that being said the user might want to know that 1.2 will be turned into 1. We could use https://github.com/rl-institut/multi-vector-simulator/blob/dev/docs/MVS_parameters_list.csv for enforcing the right type of parameter)

smartie2076 commented on src/multi_vector_simulator/utils/data_parser.py:
```diff
@@ -283,20 +287,44 @@ def convert_epa_params_to_mvs(epa_dict):
+        if len(d.keys()) > 0:
+
+            for asset_group in d.keys():
+                errror_msg.append(asset_group)
```
`errror` with three `r`

Bachibouzouk commented:
@smartie2076 - I pulled ´backup/adapt-epa-parser_martha´ locally, I still had the previous version of `fix/adapt-epa-parser` locally, so I added your commits with ´git cherry-pick <your commit number>´ one by one (there is certainly a smarter way to do this), then I rebased onto latest ´dev´ and here we go. There were a few conflicts which I think I solved the right way. You'll be able to double check when reviewing.

Bachibouzouk commented on src/multi_vector_simulator/utils/data_parser.py:
```diff
@@ -214,6 +219,7 @@
         STORAGE_CAPACITY,
         "optimize_capacity",
         "input_timeseries",
+        # TODO split flow into in, out and storage
```
this will be discussed bilaterally with EPA today

smartie2076 commented:
One remark: With #184 and #730 there will be two parameters added to the input files. We need to remember to add them to the parser. I think that in connection with #750 we should think of using the default parameters also for the parser when we implement it...

Bachibouzouk commented:
This is the file which is provided by EPA benchmark model
[json2MVS.json.txt](https://github.com/rl-institut/multi-vector-simulator/files/5789625/json2MVS.json.txt)
With the parser in its current state it leads to the following error and warnings:

```
ERROR:root:The age of the asset `demand_01` (10 years) is lower or equal than the asset lifetime (10 years). This does not make sense, as a replacement is imminent or should already have happened. Please check this value.
WARNING:root:You might use an old input file! The efficiency of the storage capacity of 'storage' is 0.001, although it should represent the ability of the storage to hold charge over time; check PR #676.
ERROR:root:Energy system bus Electricity (DSO)_pdp has too few assets connected to it. The minimal number of assets that need to be connected so that the bus is not a dead end should be two, excluding the excess sink. These are the connected assets: Electricity_grid_DSO_consumption


WARNING:root:The calculation of the Total_excess per sector and their energy equivalent may currently be faulty. Please refer to issue #559
WARNING:root:Attention, on bus Electricity there is excessive excess generation, totalling up to 18% of the inflows. The total inflows are 105113 and outflows 86049  It seems to be cheaper to have this excess generation than to install more capacities that forward the energy carrier to other busses (if those assets can be optimized).

```

smartie2076 commented:
> ERROR:root:The age of the asset `demand_01` (10 years) is lower or equal than the asset lifetime (10 years). This does not make sense, as a replacement is imminent or should already have happened. Please check this value.

Energy consumption assets usually not use lifetime or an age. Both parameters shouldn't exist for this asset types (others that should not be there compage [above, point 15](https://github.com/rl-institut/multi-vector-simulator/pull/675#issuecomment-742656073). Their default value should be zero. 

> WARNING:root:You might use an old input file! The efficiency of the storage capacity of 'storage' is 0.001, although it should represent the ability of the storage to hold charge over time; check PR #676.

This should tell you to check the efficiency of your storage. If the efficiency of storages is only 0.01, it means that 99% of your stored energy is lost within one time step (self-discharge). So you have to adapt the value in your input file (from EPA). I dont think there is anything wrong with the EPA in this though.

> ERROR:root:Energy system bus Electricity (DSO)_pdp has too few assets connected to it. The minimal number of assets that need to be connected so that the bus is not a dead end should be two, excluding the excess sink. These are the connected assets: Electricity_grid_DSO_consumption

This is fine, and always occurs. the is already issue about it, but basically you can ignore it.

> WARNING:root:The calculation of the Total_excess per sector and their energy equivalent may currently be faulty. Please refer to issue #559

This is an open issue that we still have to fix, and then the warning will also go away. the user input has nothing to do with this.

> WARNING:root:Attention, on bus Electricity there is excessive excess generation, totalling up to 18% of the inflows. The total inflows are 105113 and outflows 86049  It seems to be cheaper to have this excess generation than to install more capacities that forward the energy carrier to other busses (if those assets can be optimized).

This is a valid warning for the user, as it might hint towards weird outputs. also tells the user what should be changed to influence the weird behaviour.

Bachibouzouk commented on src/multi_vector_simulator/utils/data_parser.py:
```diff
@@ -239,9 +277,23 @@ def convert_epa_params_to_mvs(epa_dict):
+                        UNIT: "minutes",
+                        VALUE: timestep,
+                    }
+            # TODO uncomment following lines when PR #722 is merged
```
Linked to #722

Bachibouzouk commented:
> This should tell you to check the efficiency of your storage. If the efficiency of storages is only 0.01, it means that 99% of your stored energy is lost within one time step (self-discharge). So you have to adapt the value in your input file (from EPA). I dont think there is anything wrong with the EPA in this though.

The value in the EPA file (the json2MVS.json) for "capacity" is 0.001

smartie2076 commented:
> The value in the EPA file (the json2MVS.json) for "capacity" is 0.001

You mean this, right?

```

  "energy_storage": [
    { 
      "capacity": {
        "efficiency": {
          "unit": "factor",
          "value": 0.001
        },
      }
    }
  ],
```

We need to tell them to change it in their benchmark test, as it is not correct anymore. It should also not be in our benchmark test in the MVS anymore.

Bachibouzouk commented:
@smartie2076 - I'll wait for your review before I rebase as `dev` was updated and there is now a conflict with the changelog :)

smartie2076 commented on src/multi_vector_simulator/utils/data_parser.py:
```diff
@@ -331,6 +358,21 @@ def convert_epa_params_to_mvs(epa_dict):
+                # TODO remove this when change has been made on EPA side
+                if asset_group == ENERGY_STORAGE:
+
+                    if OPTIMIZE_CAP not in dict_asset[asset_label]:
+                        dict_asset[asset_label][OPTIMIZE_CAP] = {
+                            "unit": "bool",
+                            "value": False,
+                        }
```
See above, I thought that when the sub-assets all (or, for that matter, any) have `OPTIMIZE_CAP=True` we should set this to True?
Is `OPTIMIZE_CAP` usually in the json from EPA right now?

You could use the constants from `constants_json_strings.py` here.

smartie2076 commented on src/multi_vector_simulator/utils/data_parser.py:
```diff
@@ -373,6 +383,13 @@ def convert_epa_params_to_mvs(epa_dict):
                         )
+
+                if asset_group == ENERGY_CONSUMPTION:
+                    if DSM not in dict_asset[asset_label]:
```
`TYPE_ASSET` does not need the default value? For as long as we have not removed them properly? (#376)

smartie2076 commented on src/multi_vector_simulator/utils/data_parser.py:
```diff
@@ -373,6 +383,13 @@ def convert_epa_params_to_mvs(epa_dict):
+                    if DSM not in dict_asset[asset_label]:
+                        dict_asset[asset_label][DSM] = False
+
+                if EMISSION_FACTOR not in dict_asset[asset_label]:
+                    dict_asset[asset_label][EMISSION_FACTOR] = {VALUE: 0}
```
Shouldn´t this be automatically be updated with the default value - or do you want to avoid the warning message?

Would it be possible to use the default value as provided in `constants.py`?

Bachibouzouk replied:
The idea was to prevent not understanding why the EMISSION_FACTOR value is always 0 if it is implemented in the EPA in the future ;)

Using the default value is a good idea :)

Bachibouzouk replied:
Done in 34dd06282c6da8e6395626b7c4730c2a89a3e63c

smartie2076 commented on src/multi_vector_simulator/utils/data_parser.py:
```diff
@@ -512,6 +512,29 @@ def convert_mvs_params_to_epa(mvs_dict, verbatim=False):
                 if asset_group == ENERGY_BUSSES and k == "Asset_list":
                     asset["assets"] = list(asset.pop(k).keys())
+                if asset_group == ENERGY_STORAGE:
+                    # if k in (MAP_MVS_EPA[INPUT_POWER], MAP_MVS_EPA[OUTPUT_POWER], MAP_MVS_EPA[STORAGE_CAPACITY]):
```
What does this commented line mean?

Bachibouzouk replied:
I forgot to remove it

smartie2076 commented on src/multi_vector_simulator/utils/constants.py:
```diff
@@ -88,7 +88,7 @@
 # associated to each of these filenames
 REQUIRED_CSV_PARAMETERS = {
-    CONSTRAINTS: [MINIMAL_RENEWABLE_FACTOR],
+    CONSTRAINTS: [MINIMAL_RENEWABLE_FACTOR, MAXIMUM_EMISSIONS],
```
I would have thought that we keep the ´MAXIMUM_EMISSIONS` as extra parameters with default values? Why not?

Bachibouzouk replied:
Every required parameters listed in `KNOWN_EXTRA_PARAMETERS` will have the default value if not provided

smartie2076 commented on src/multi_vector_simulator/utils/data_parser.py:
```diff
@@ -390,10 +392,11 @@ def convert_epa_params_to_mvs(epa_dict):
-            logging.warning(
-                f"The parameters {MAP_MVS_EPA[asset_group]} are not present in the parameters to be parsed into mvs json format"
+            logging.INFO(
+                f"The assets parameters '{MAP_MVS_EPA[asset_group]}' is not present in the EPA parameters to be parsed into MVS json format"
```
Does this only parse the asset group or also the asset name?

Bachibouzouk replied:
This message is only there if the whole asset group is missing

smartie2076 reviewed (approved):
Puh, that looks like it was a lot of work! Nice! =D

I think you can check again which issues you close with this PR (#3? #605? #606?).

For all comments that are minor/cosmetic you can also merge this PR and create a new one to fix those.

Bachibouzouk commented on src/multi_vector_simulator/D0_modelling_and_optimization.py:
```diff
@@ -101,8 +101,8 @@ def run_oemof(dict_values, save_energy_system_graph=False):
     )
-
-    model_building.store_lp_file(dict_values, local_energy_system)
+    if save_lp_file is True:
+        model_building.store_lp_file(dict_values, local_energy_system)
```
Yes maybe that is a better option, you are right :=)

Bachibouzouk replied:
Done in a5f6083c58bebb2307a738316c586376935bda3e

Bachibouzouk commented on src/multi_vector_simulator/utils/data_parser.py:
```diff
@@ -318,6 +318,33 @@ def convert_epa_params_to_mvs(epa_dict):
+
+                            # remove non-implemented parameter if provided faultily
+                            if OPTIMIZE_CAP in subasset:
+                                subasset.pop(OPTIMIZE_CAP)
```
This is dealt with lower in the file, this removes `OPTIMIZE_CAP` for subassets of storage asset

Bachibouzouk commented on src/multi_vector_simulator/utils/data_parser.py:
```diff
@@ -318,6 +318,33 @@ def convert_epa_params_to_mvs(epa_dict):
+                                subasset.pop(OPTIMIZE_CAP)
+
+                            # add unit if not provided
+                            if UNIT not in subasset:
+                                if k == MAP_MVS_EPA[STORAGE_CAPACITY]:
+                                    subasset[UNIT] = "kWh"
+                                else:
+                                    subasset[UNIT] = "kW"
```
I will add a TODO

SabineHaas commented on src/multi_vector_simulator/D0_modelling_and_optimization.py:
```diff
@@ -101,8 +101,7 @@ def run_oemof(dict_values, save_energy_system_graph=False, save_lp_file=False):
     )
-    if save_lp_file is True:
-        model_building.store_lp_file(dict_values, local_energy_system)
+    model_building.store_lp_file(dict_values, local_energy_system)
```
@Bachibouzouk, in the changelog you wrote 
"Write lp file only when executing `cli.py` (#675)"
, but here it looks like it will be stored for every simulation. Could you clarify this please?

Bachibouzouk replied:
I forgot to add "if user set the `OUTPUT_LP_FILE` setting to `True`"
